docs/programmers_guide/witness_formal_spec.md

```diff
@@ -0,0 +1,222 @@
+STACK(helper1(node1, node2), helper2(hash1, hash2))
+```
+
+This substitution rule replaces 2 stack items `(node1, hash1)` and `(node2, hash2)`
+with a single stack item `(helper1(node1, node2), helper2(hash1, hash2))`.
+
+Where `helper1` and `helper2` are pure functions defined in pseudocode (see the example below).
+
+```
+helper1 (value1, value2) {
+    return value1 + value2
+}
+
+helper2 (hash1, hash2) {
+    return KECCAK(hash1 + hash2)
+}
+```
+
+The specification for a substitution rule
+
+```
+[GUARD <CONDITION> ...]
+
+[ STACK(<node-var-name>, <hash-var-name>), ... ] <INSTRUCTION>[(<params>)] |=>
+STACK(node-value, hash-value), ...
+```
+
+The substitution rule MAY have one or more GUARD statements.
+The substitution rule MAY have one or more STACK statements before the instruction.
+The substitution rule MUST have exactly one instruction.
+The substitution rule MAY have parameters for the instruction.
+The substitution rule MUST have at least one STACK statement after the arrow.
+
+So, the minimal substitution rule is for the `HASH` instruction that pushes one hash to the stack:
+```
+HASH(hashValue) |=> STACK(hashNode(hashValue), hashValue)
+```
+
+## GUARDs
+
+Each substitution rule can have zero, one or multiple `GUARD` statements.
+Each `GUARD` statement looks like this:
+
+```
+GUARD <CONDITION>
+```
+
+That means that for the substitution rule to be applicable the `<CONDITION>` in the guard statement must be true.
+
+If a substitution rule has multiple guard statements, all of the conditions specified there should be satisfied.
+
+## `TRAP` statements
+
+If no substitution rules are applicable for an instruction, then the execution MUST stop (`TRAP` instruction) and the partial results MUST NOT be used.
+
+Implementation of the `TRAP` instruction is undefined, but it should stop the execution flow. (e.g. in Golang it might be a function returning an error or a panic).
+
+## Instructions & Parameters
+
+Each instruction MAY have one or more parameters.
+The parameters values MUST be situated in the witness.
+The parameter values MUST NOT be taken from the stack.
+
+That makes it different from the helper function parameters that MAY come from the stack or MAY come from the witness.
+
+## Helper functions
+
+Helper functions are functions that are used in GUARDs or substitution rules.
+
+Helper functions MUST be pure.
+Helper functions MUST have at least one argument.
+Helper functions MAY have variadic parameters: `HELPER_EXAMPLE(arg1, arg2, list...)`.
+Helper functions MAY contain recursion.
+
+## Data types
+
+INTEGER - we treat integers as infinite, the overflow behaviour or mapping to the actual data types is undefined in this spec and should be dependent on implementation.
+
+## Execution flow 
+
+Let's look at the example.
+
+Our example witness would be `HASH h1; HASH h2; BRANCH 0b11`.
+
+Initial state of the stack is ` <empty> `;
+
+---
+
+**1. Executing `HASH h1`**: push a `hashNode` to the stack.
```

Comment:
What is a hash node?

Reply:
I added some explanation and a list of possible nodes earlier in the doc.